### ISSUE TITLE
[QUESTION] Include possible HTTPExceptions in OpenAPI spec

### ISSUE THREAD

**Issue by onecrayon** [tiangolo/fastapi#1999]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi import FastAPI, HTTPException

app = FastAPI()


@app.get("/")
def read_root(gimme_coffee: bool = False):
    if gimme_coffee:
        raise HTTPException(status_code=418, detail="I'm a teapot.")
    return {"Hello": "World"}
```

### Description

* Open the browser and call the endpoint `/docs`
* Note that the only listed response codes are 200 and 422
* Execute the route with `gimme_coffee` set to true and note that it returns a 418 status code

How do I set the status code for 418 in the OpenAPI docs? I was expecting to be able to pass in a list of possible HTTPExceptions and have them be automatically converted, but there does not appear to be any way to do that. The closest thing I found was [Additional Status Codes](https://fastapi.tiangolo.com/advanced/additional-status-codes/) but that seemed at best tangentially relevant since I am raising HTTPExceptions, not looking to build out a custom exception JSON response.

Properly documenting error states is just as important as documenting the success state. I really hope FastAPI provides an easy way to do this.

### Environment

* OS: Linux (Docker container using standard python-3.8 image)
* FastAPI Version: 0.61.1
* Python version: 3.8.5

**Comment by ycd** [comment]
Yes, you can declare additional responses with additional status codes etc. You just need to add it as a path parameter then those additional responses will be included in the OpenAPI schema, here is a example

```python
@app.get("/",responses={418: {"detail": "I'm a teapot"}, 413: {"Too large": "...The Payload"}})
def read_root(gimme_coffee: bool = False):
    if gimme_coffee:
        raise HTTPException(status_code=418, detail="I'm a teapot.")
    return {"Hello": "World"}
```

Documentation for  [Additional Responses](https://fastapi.tiangolo.com/advanced/additional-responses/)

**Comment by onecrayon** [comment]
That's the additional status codes I mentioned above, and it is a long way from easy. I was hoping I could instantiate an error, pass it in similar to `response_model` and have the OpenAPI generation just do the right thing, but looks like I would need to basically duplicate the entire error into the decorator.

Nuts. I guess this is actually a feature request: we need a better way to handle error message documentation. My preference would be to just pass in a list of some base error classes and have it figure out the status code and output formatting based on that (having the actual error output isn't necessary, particularly since it may well change depending on the situation). That way, I could setup a sub-class of HTTPException that defaults to a 418 status code, and by passing that in and using it my exception is fully documented.

**Comment by David-Lor** [comment]
@onecrayon On my APIs I'm using custom exceptions that include references to the error code, message and Pydantic model to be returned. I have an example over here: https://github.com/David-Lor/FastAPI-Pydantic-Mongo_Sample_CRUD_API

On a nutshell:

- I create the custom exceptions (https://github.com/David-Lor/FastAPI-Pydantic-Mongo_Sample_CRUD_API/blob/master/people_api/exceptions.py)
- I create the models for exception responses (https://github.com/David-Lor/FastAPI-Pydantic-Mongo_Sample_CRUD_API/blob/master/people_api/models/errors.py)
- On the routes, I call a function that creates the "responses" dict, given multiple custom exceptions (https://github.com/David-Lor/FastAPI-Pydantic-Mongo_Sample_CRUD_API/blob/master/people_api/app.py)
- I use a request middleware, where a try/except can catch my custom exceptions, and return their custom response, based on the exception status code, message and model (https://github.com/David-Lor/FastAPI-Pydantic-Mongo_Sample_CRUD_API/blob/master/people_api/middlewares.py)

Probably far from ideal and might be better ways to achieve it, but for now it's been working pretty fine for me.

**Comment by ycd** [comment]
@onecrayon Does something like this would work for you?`

```python
from fastapi import FastAPI, APIRouter


app = FastAPI()
router = APIRouter()


@router.get("/dummy1")
async def dummy1():
    ...


@router.get("/dummy2")
async def dummy2():
    ...


@router.get("/dummy3")
async def dummy3():
    ...


@router.get("/dummy4")
async def dummy4():
    ...


app.include_router(
    router,
    responses={
        418: {
            "description": "I'm a teapot",
            "content": {"application/json": {"example": {"dummy": "dumdum"}}},
        },
        413: {
            "description": "I'm too large",
            "content": {"application/json": {"example": {"dummy": "dumdum"}}},
        },
    },
)
```



![image](https://user-images.githubusercontent.com/62724709/100586198-d6febc80-32ff-11eb-9bae-37d2500b2039.png)

**Comment by onecrayon** [comment]
@ycd Per [my comment here](https://github.com/tiangolo/fastapi/issues/1999#issuecomment-686228362), that's what I'm doing. It's also unnecessary code duplication. My feature request is for an easy way to pass in children of HTTPException and just have FastAPI automatically figure out what the error output should look like. So for example here's some pseudo-code that's based on my current project:

```python
from fastapi import status, HTTPException

class APIException(HTTPException):
    """Light wrapper around HTTPException that allows specifying defaults via class property"""

    status_code = status.HTTP_400_BAD_REQUEST
    detail = None
    headers = None

    def __init__(self, *args, **kwargs):
        if "status_code" not in kwargs:
            kwargs["status_code"] = self.status_code
        if "detail" not in kwargs:
            kwargs["detail"] = self.detail
        if "headers" not in kwargs:
            kwargs["headers"] = self.headers
        super().__init__(*args, **kwargs)

class TeapotException(APIException):
    status_code = 418
    detail = "I'm a teapot."

@app.get("/")
def read_root(gimme_coffee: bool = False, response_exceptions=[TeapotException]):
    if gimme_coffee:
        raise TeapotException()
    return {"Hello": "World"}
```

Even better would be to have FastAPI able to introspect the method and look for HTTPException-derived exceptions to auto-populate the list, but that's possibly edging a little too far into magical territory.

In any case, it's silly that I can have FastAPI automatically discover query params and such based on passed classes, but need to explicitly duplicate all of the information that's already stored in my exception classes in order to document exceptions.

**Comment by skaanbilly2** [comment]
Is this issue on the planning, or can I create a PR for this?

**Comment by Kludex** [comment]
Does this helps? https://github.com/Kludex/fastapi-responses

**Comment by onecrayon** [comment]
@Kludex That is *exactly* the sort of thing that I'm looking for! Granted, it's a little more "magical" than I usually like, but I would happily take it over the current setup where I have to explicitly define status codes and error strings multiple times.

**Comment by jtv8** [comment]
@onecrayon - given that the main use case for this is setting some expectations for API users on what HTTP responses to handle, would this suffice?

```
@app.get("/", response_exceptions=[418])
def read_root(gimme_coffee: bool = False):
    if gimme_coffee:
        raise HTTPException(status_code=418, detail="I'm a teapot.")
    return {"Hello": "World"}
```

**Comment by onecrayon** [comment]
@jtv8 No, that's insufficient. The API spec needs to be able to output not only the status code, but also the specification for the response body (since different exceptions often have different JSON body responses that get rendered, or other methods for reporting what went wrong).

**Comment by thedmdim** [comment]
Any update here?
